shakenfist/etcd.py

```diff
@@ -31,19 +31,21 @@ class ReadException(Exception):
     pass
 
 
-def get_lock(name, ttl=60):
-    start = time.time()
-    for attempt in range(ETCD_ATTEMPTS):
-        try:
-            return etcd3.client().lock(name, ttl=ttl)
-        except Exception as e:
-            LOG.info('Failed to acquire lock, attempt %d: %s' % (attempt, e))
-            time.sleep(ETCD_ATTEMPT_DELAY)
-        finally:
-            LOG.debug('Locked etcd key "%s" after %.02f seconds'
-                      % (name, time.time() - start))
+class ActualLock(etcd3.Lock):
+    def __init__(self, name, ttl=60, etcd_client=None, timeout=10):
+        super(ActualLock, self).__init__(name, ttl, etcd_client)
+        self.timeout = timeout
+
+    def __enter__(self):
+        LOG.debug("ActualLock.__enter__() timeout=%s name=%s",
+                  self.timeout, self.name)
+        if not self.acquire(timeout=self.timeout):
+            raise LockException('Cannot acquire lock: %s' % self.name)
+
```

Comment:
The original enter method returns self here. I feel like that might be important?

Reply:
Additionally, should we override __exit__() as well and perhaps log how long the lock was held for or something?

Reply:
The exit() does the correct release() operation. No need to change. We are just fixing the "broken" enter(). 
The original enter() naively calls acquire() without specifying a timeout of zero. Everybody assumes that "with" would not enter the code block without acquiring the lock.

Adding timing is obviously simple but would you want to elevate it above debug? Personally I think that would make the logs unnecessarily noisy.

No prob to add it, if you want.

Reply:
What about __enter__() returning self? That's used in patterns like:

```
with get_lock('banana') as l:
  ...
  l.refresh()
  ...
```

So I think you do need to return self there?

Reply:
Absolutely correct. I overlooked your first comment. Fixed.

Reply:
You still don't return self here?

Reply:
So it turns out that I don't know how to use git....